File: DWITE 2012 round 1/q3.hs
import Data.List
parseInput :: String -> [[String]]
parseInput = map (words) . lines

compute d = reverse . snd $ foldl person ([],[]) d 
  where
    newConflicts costume costumes conflicts name =
      maybe conflicts (\_ -> name:conflicts) $ find (costume==) costumes
    person (costumes,conflicts) [name,costume] = 
      (costume : costumes,newConflicts costume costumes conflicts name)

answer [] = []
answer ([n]:rest) = (compute d) : (answer nextData)
  where
    num = (read n) :: Int
    (d,nextData) = splitAt num rest

main = do
  inputData <- readFile "DATA3.txt"
  let input = parseInput inputData
      answers = answer input
      answers' = unlines . map (\x -> if null x then "SPOOKY" else unwords x) $ answers
  writeFile "OUT3.txt" answers'
  

File: DWITE 2012 round 1/q3.rb
debug = false

input = debug ? STDIN : File.new("DATA3.txt","r")
out = debug ? STDOUT : File.open("OUT3.txt", 'w')

5.times do
  n = input.gets.to_i
  used = {}
  collisions = []
  n.times do
  	line = input.gets.split(" ")
  	costume = line[1]
  	if used[costume]
  		collisions << line[0]
  	else
  		used[costume] = true
  	end
  end
  if collisions.length > 0
  	out.puts collisions.join(" ")
  else
  	out.puts "SPOOKY"
  end
end